refnx/dataset/test/test_reflectdataset.py

```diff
@@ -261,25 +273,25 @@ def test_mask(self):
         assert_equal(len(self.data), self.data._y.size)
 
     def test_repr(self):
-        a = Data1D(os.path.join(self.pth, "c_PLP0033831.txt"))
+        a = Data1D(self.pth / "c_PLP0033831.txt")
         b = eval(repr(a))
         assert_equal(len(a), 166)
         assert_equal(len(b), 166)
 
         # load dataset from XML, via string
-        a = ReflectDataset(os.path.join(self.pth, "c_PLP0000708.xml"))
+        a = ReflectDataset(str(self.pth / "c_PLP0000708.xml"))
         b = eval(repr(a))
         assert_equal(len(b), len(a))
         assert_equal(len(b), len(a))
 
 
 class TestData1D:
     @pytest.fixture(autouse=True)
-    def setup_method(self, tmpdir):
-        self.pth = os.path.dirname(os.path.abspath(__file__))
-        self.cwd = os.getcwd()
-        self.tmpdir = tmpdir.strpath
-        os.chdir(self.tmpdir)
+    def setup_method(self, tmp_path):
+        self.pth = Path(__file__).absolute().parent
+        self.cwd = Path(".").resolve()
```

Comment:
`Path.cwd()`; no idea how portable `.` is as cwd.  (There are lots of other places where `Path(".")` was used; I wasn't sure if some of them should be based on `__file__` rather than cwd)

A problem for another day: this doesn't actually look like a pytest [fixture](https://docs.pytest.org/en/latest/how-to/fixtures.html), rather compatibility layer to nose/xunit. Combining the setup/teardown methods into one, yielding in the middle, and adding that function to the test methods that need it would be the pytest fixture idiom.

Reply:
I assume `tmp_path` is a fixture, https://docs.pytest.org/en/7.1.x/how-to/tmp_path.html#the-tmp-path-fixture

Reply:
Yes it is, but the `pytest.fixture` decorator is for when you create new fixtures, not for when you use fixtures.

There's no need for a decorator there to cause the `tmp_path` fixture to be used, that's done behind the scenes by pytest itself. pytest sees `tmp_path` in the method signature and invokes the fixture for it, feeding in the result.

If a fixture is specifically requested by including it in the signature of the methods that use it, then that fixture just needs `@pytest.fixture`.  If you want a fixture to be run all the time without including it in the signatures of each method, then the `autouse=True` parameter arranges that within that scope. 

In this case, decorating `setup_method` as an autouse fixture makes it like a fixture for all test methods within the class, but it's not really written like a fixture - it's written like an old-school pair of setup/teardown methods. If that decoration were actually needed, then I suspect the cleanup code is never run (because there's no end-of-fixture hook like that; that's what a yield fixture is for). 

Rather than this actually being a fixture, this could be run with one of pytest's other compatibility shims that exist to help convert test suites from other testing approaches. The setup_method/teardown_method idiom is one of them, and fixture marking isn't needed for that. Similarly, pytest also has `unittest` compatibility that lets you inherit from `unitest.TestCase` and have `setUp` methods (but fixtures don't work with that very well). 

Probably, however, you want this to be a `chdir_tmp_path` fixture that requests the `tmp_path` fixture, chdirs to it, yields, then chdirs back. Some error handling on that is easier, `self` is actually unneeded for that too.

Reply:
when I get rid of all the decorators nothing works. I don't understand what the correct thing is to do here.

Reply:
`setup_method` currently needs to be annotated because it doesn't have the method signature signature that is required for the xunit compat shim to kick in. `tear_down` does use that compat shim because it has the right signature. That works, but is unusual, and one has to remember why it works for when it breaks one day :)

For TestData1D, the following is more idiomatic:
```python
    @pytest.fixture(autouse=True)
    def tmp_path_chdir(self, tmp_path):
        cwd = Path.cwd()
        os.chdir(tmp_path)

        try:
            yield
        finally:
            os.chdir(cwd)
```

It is possible to go one step further than that and make it an explicit use fixture rather than an autouse fixture; that only involves adding `tmp_path_chdir` to the function definitions of each of the functions that need it. Given how often this particular sequence is used (11 times in the test suite, it seems), it might be worth making it an explicitly used fixture and adding it to `conftest.py` so that it's not repeated multiple times.

Converting the `setup_method` in `TestReflectDataset` could be done by splitting it into three - each fixture only does one thing, and no side effects:
```python
    @pytest.fixture(autouse=True)
    def tmp_path_chdir(self, tmp_path):
        cwd = Path.cwd()
        os.chdir(tmp_path)

        try:
            yield
        finally:
            os.chdir(cwd)

    @pytest.fixture
    def dummy_data(self):
        data = ReflectDataset()

        x1 = np.linspace(0, 10, 115)
        y1 = 2 * x1
        e1 = np.ones_like(x1)
        dx1 = np.ones_like(x1)
        data.add_data((x1, y1, e1, dx1))
        yield data

    @pytest.fixture
    def data_location(self):
        location = Path(__file__).absolute().parent
        yield location

    def test_load(self, data_location):
        # load dataset from XML, via file handle
        dataset = ReflectDataset()
        with open(data_location / "c_PLP0000708.xml") as f:
            dataset.load(f)
```

`unittest` is very focused on classes providing scope for setup/teardown methods, it encourages the use of `self` throughout the tests to store parameters and data related to the harness. The setup/teardown interaction with the test functions relies on side-effects by storing state in `self`, but that also reduces isolation between test functions and so tests can accidentally influence each other. `pytest` instead manages it all via fixtures and the use of `self` is normally not needed or wanted. With `pytest`, classes are (more or less) only used for an additional level of scoping to permit easy selection of sets of tests on the commandline, for applying marks, and sometimes for autouse fixtures

(I have this particular file all updated to this more idiomatic fixture use if you want a PR for it; there's no need to update all at once, it can be a "when next touching these tests" process.)